Add country data to {daedalus}

pratikunterwegs opened the pull request:
This PR adds data from the P2 Drivers repository to {daedalus}:
1. `country_data`: A list of country-wise data on demography, age-stratified contact matrices, and number of workers in each economic sector;
2. `country_names`: A list of country names for user reference.
3. `economic_contacts`: A list of worker-to-worker contacts within and between sectors. Contacts between sectors currently take a dummy value.

Other changes:
1. The function `daedalus()` can now be run with a single `country` argument, which automatically pulls country data from the package to initialise epidemiological compartments and economic sector data. Users can still pass their own data (e.g. a different contact matrix or economic sector worker distribution), as well as assumptions about initial infections, as named arguments via `...`;
2. Function `default_parameters()` has been pulled internal;
3. Internal functions `make_initial_state()` and `make_parameters()` added to help with 
4. Model constants have been pulled internal;
5. Readmes and the 'Get started' vignette have been updated.
6. Documentation and tests have been updated.

This PR is an intermediate step before a few more data packages are added and should be a good way to see whether we're happy with adding, providing, and accessing data in this way. The data pre-processing in `data-raw/` still needs to be confirmed with EPPI (scheduled), so there might be tweaks to the format.

Some incoming data that haven't been added:
1. Pathogen scenario data - data in P2 Drivers is ambiguous and needs clarification;
2. Intervention scenario data will be added along with/after intervention triggers are added.

richfitz commented on tests/testthat/test-daedalus.R:
```diff
@@ -1,13 +1,11 @@
   )
 
-  output <- do.call(daedalus, default_inputs())
+  output <- daedalus("Canada")
```
You can save this within the `expect_no_condition` on l5 if you prefer:

```
expect_no_condition(
  output <- daedalus("Canada")
  )
```

richfitz commented on tests/testthat/test-daedalus.R:
```diff
@@ -78,64 +73,23 @@ test_that("daedalus: statistical correctness", {
-  expect_error(
-    daedalus(initial_state_),
-    regexp = "(Expected `initial_state`)*(array)*(with no missing elements)"
+    daedalus("U.K."),
```
This sort of abbreviation (and the inconsistencies in country name - why not "United Kingdom of Great Britain and Northern Ireland" which is the full name) is why we end up going with ISO3 often

richfitz reviewed (changes requested):
Thanks - I think the approach here is great (in fact the cowflu package has an almost identical construction for processing raw data).

Things to look at:

* the data are missing but there's no docs (or I missed them sorry) on how to get the data in to run the processing scripts
* I'm not sure that colloquial english-localised country name is the best selector
* the regular expressions in the tests are not behaving the way I think you think they are

Otherwise looking great. As I mention in a comment we've not usually used the `data()` approach in other packages, but it looks like a nice fit here

pratikunterwegs commented on R/country_data.R:
```diff
@@ -0,0 +1,35 @@
+#' @format ## `country_data`
+#' A list with 197 elements, each corresponding to a recognised country or
+#' territory.
+#' \describe{
```
That's an interesting question I've never considered - I can check, but suspect not otherwise [this example in R Packages](https://r-pkgs.org/data.html#sec-documenting-data) would say that?

pratikunterwegs replied:
The documentation does accept an `md` list (left) but loses the IMO slightly nicer LaTeX-like description formatting (right). I can replace if it seems easier but at the moment there's little cost, maintenance-wise.
![Screenshot from 2024-08-01 15-21-43](https://github.com/user-attachments/assets/b5a552c3-67c4-44b8-882e-4a4259e6f11c)

richfitz replied:
yeah, your version looks much nicer. It's a pity that there's no tweak on the md version to allow this

pratikunterwegs commented on R/daedalus.R:
```diff
@@ -1,59 +1,114 @@
-#' @param initial_state The initial state of the population as a numeric array
-#' where rows represent age groups, columns represent epidemiological
-#' compartments, and layers represent economic sectors.
+#' @param country A country or territory name from among `COUNTRY_NAMES`.
```
Yep, happy to modify this. I can ask EPPI what they might want as well.

pratikunterwegs replied:
EPPI are fine with the ISO-3 codes but would also like a dictionary included mapping names to codes. There's a small wrinkle where the Channel Islands are included in the data but have no code (understandably).

richfitz replied:
We could follow https://en.wikipedia.org/wiki/XK_(user_assigned_code) and assign them to XUC or something else starting with X?

pratikunterwegs replied:
Sounds good - I'll add that to the task list for the next PR.

pratikunterwegs commented on R/daedalus.R:
```diff
@@ -1,59 +1,114 @@
+#' `make_parameters()` allows users to pass the following pathogen parameters:
+#'
+#' - `beta`: A single numeric value > 0.0 for the infection transmission rate
+#' \eqn{\beta}. Defaults to 0.0186, assuming an \eqn{R_0} of 1.3
```
Yes, seems to be a quirk I've been bitten by in the past as well. I can check if there's a workaround as this lack of support has been the case for a couple of years at least

pratikunterwegs commented on R/daedalus.R:
```diff
@@ -1,59 +1,114 @@
-      missing elements but it is {.obj_type_friendly {initial_state}} or has
-      missing elements."
+daedalus <- function(country, ...,
+                     time_end = 300L) {
```
It can, but I think {lintr} complains about implicit integers since I've turned on that linter - I can turn it off.

richfitz replied:
yeah, turning that linter off seems sensible

pratikunterwegs commented on R/helpers.R:
```diff
@@ -0,0 +1,72 @@
+  )
+
+  # get demography and sector workforce
+  demography <- daedalus::country_data[[country]][["demography"]]
```
1. Apparently yes, although I'm a little unclear on why - I've just accepted it as an R quirk.
2. I've turned the extraction linter back on, so accessing via `$` would cause it to complain.

pratikunterwegs commented on R/helpers.R:
```diff
@@ -0,0 +1,72 @@
+
+  # calculate non-working working-age, and prepare initial state accounting
+  # for distribution of working age into economic sectors
+  inactive_workers <- demography[i_WORKING_AGE] - sum(sector_workforce)
```
No, I can add that - there are definitely some data with `NA`s so it's possible negative values could creep in.

pratikunterwegs commented on README.Rmd:
```diff
@@ -37,96 +37,28 @@ remotes::install_github("j-idea/daedalus", upgrade = FALSE)
-
-# get demography vector
-demography <- contact_data[["demography"]][["population"]]
+The model can be run for any country or territory in the `COUNTRY_NAMES` list by passing the country name to the function.
```
Thanks, actually it should be `country_names` - I think once the package is loaded the data should be available without namespacing.

pratikunterwegs commented on README.Rmd:
```diff
@@ -37,96 +37,28 @@ remotes::install_github("j-idea/daedalus", upgrade = FALSE)
-# get demography vector
-demography <- contact_data[["demography"]][["population"]]
+The model can be run for any country or territory in the `COUNTRY_NAMES` list by passing the country name to the function.
+This automatically pulls country-specific demographic and economic data, which is included in the package, into the model (see the ['Get started' vignette](articles/daedalus.html) for more details).
```
Thanks, I can check this - it wouldn't work on a local {pkgdown} build without the directory.

richfitz replied:
yeah, there's no great way of having a single readme that renders well for both the github source view and the landing page for the pkgdown site.  We've solved this specific issue elsewhere by using the absolute url here (so in your case `See the ['Get started vignette](https://jameel-institute.github.io/daedalus/articles/daedalus.html)` which renders and links nicely in both contexts

pratikunterwegs commented on data-raw/country_data.R:
```diff
@@ -0,0 +1,164 @@
+# Walker et al. 2020
+
+# load country data
+country_data <- fread("data-raw/country_data.csv")
+sector_contacts <- fread("data-raw/sectorcontacts.csv")
```
Yes, that's the intent. It does make checking changes in the raw data more difficult, but package data are typically intended to be fairly static, unless it's a dedicated data package.

I'm going to figure out the source with Rob and Patrick tomorrow, and update the script to reflect that if possible (e.g. with a download from a m/s supplementary material link).

pratikunterwegs replied:
After speaking with them, the data provenance is more messy than expected. Some preprocessing steps are in the [P2 repository](https://github.com/robj411/p2_drivers), but don't work as the upstream data from a paper has changed. Other data come from almost bespoke treatments of country-specific data.

pratikunterwegs commented:
Thanks, just a few broad clarifications:

> the data are missing but there's no docs (or I missed them sorry) on how to get the data in to run the processing scripts

This needs to be worked out properly (more than linking to papers purporting to provide the data)

> I'm not sure that colloquial english-localised country name is the best selector

I can work this out as well, and gently steer them in the ISO-3 direction.

> Otherwise looking great. As I mention in a comment we've not usually used the data() approach in other packages, but it looks like a nice fit here

One minor change I'll probably make is to check with EPPI whether they want the package to also serve the data to users, and if not, [I'll stick the data into `sysdata/` which is more appropriate for purely internal use](https://r-pkgs.org/data.html#sec-data-sysdata).

pratikunterwegs commented on R/daedalus.R:
```diff
@@ -64,22 +119,40 @@ daedalus <- function(initial_state, time_end = 300L,
   }
+
+  # capture parameters, check for numeric input, and split for lower-level fns
+  parameters <- list(...)
```
The list is later split up and processed by `make_parameters()`, and `make_initial_state()`. `make_initial_state()` takes only the appropriate parameters. `make_parameters()` checks for input names and currently throws a warning if any extra or mis-named arguments are passed.

I can bump that up to an error if that makes more sense (and it might if the cost of running the model before the user sees the warning is high enough; e.g. a more complex model or more timesteps).

pratikunterwegs commented on vignettes/daedalus.Rmd:
````diff
@@ -14,98 +16,28 @@ knitr::opts_chunk$set(
 ```
 
-### Run model
+The model runs for 300 timesteps (assumed to be days) by default.
````
Added in Readmes and vignette text.

pratikunterwegs commented on tests/testthat/test-helpers.R:
```diff
@@ -0,0 +1,36 @@
+  )
+  expect_error(
+    daedalus("Canada", p_infectious = -1.0),
+    regexp = "(`p_infectious` must be )*(in the range \\[0.0, 1.0\\].)"
```
I _think_ these regexes have been fixed in https://github.com/jameel-institute/daedalus/pull/11/commits/928872d33b46bcff23e3c4c46a5e7ff3785b3d82

pratikunterwegs commented on tests/testthat/test-parameters.R:
```diff
@@ -0,0 +1,168 @@
+  )
+  expect_error(
+    daedalus("Canada", beta = c(0.1, 0.1)),
+    regexp = "(Expected each parameter)*(`...`)*(to be a single)*(number)"
```
Again I think these have been fixed in https://github.com/jameel-institute/daedalus/pull/11/commits/928872d33b46bcff23e3c4c46a5e7ff3785b3d82

pratikunterwegs commented on data-raw/country_data.R:
```diff
@@ -0,0 +1,164 @@
+## code to prepare `country_data` ##
```
Added in https://github.com/jameel-institute/daedalus/pull/11/commits/96e5135d75c254d71206f1c944d1b1b98479d37f, but note one script will error as checks on two countries fail - I'm letting this pass for now while we work out the data sources and what to do about this.

pratikunterwegs commented on data-raw/country_data.R:
```diff
@@ -0,0 +1,164 @@
+## code to prepare `country_data` ##
+
+library(data.table)
```
They don't have any preferences on this.

pratikunterwegs commented:
Some final changes:
1. I've added the raw data used in `data-raw`. I'm keeping them in `data/` rather than `sysdata/` and making them available to users for now, via `daedalus::country_data` etc.
2. I've fixed the regexes to be as simple as possible.
3. The country name issue is still open, as we need a code for the Channel Islands, or to remove them or include them in the UK.
